### PR TITLE
Provide SafeDNS provider integration 

### PR DESCRIPTION
**Pull request by miff2000**
This PR is to add in support for SafeDNS, UKFast's DNS service.

Let me know if I've missed anything :)

### CODE REVIEW

**Comment by CapsuleCD** [comment]
Hi.

I'm an automated pull request bot named [CapsuleCD](https://www.capsulecd.com). I handle testing, versioning and package releases for this project. 
- If you're the owner of this repo, you can click the button below to kick off the CapsuleCD build pipeline and create an automated release.'
- If not, don't worry, someone will be by shortly to check on your pull request. 

[CapsuleCD](https://www.capsulecd.com/project/github/AnalogJ/lexicon/pullrequests/411)

---
If you're interested in learning more about [CapsuleCD](http://www.github.com/AnalogJ/capsulecd), or adding it to your project, you can check it out [here](https://www.capsulecd.com)

**Comment by AnalogJ** [comment]
Hey @miff2000 thanks for the PR! 🎉 

@adferrand and I will take a look, and get back to you if we find anything that needs to be changed. If not you're PR should be included in the next release.

File: lexicon/providers/safedns.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            'name': self._full_name(name),
+            'type': rtype,
+            'content': content
+        }
+
+        self._post('/zones/{0}/records'.format(self.domain), data)
+        LOGGER.debug('create_record: %s', True)
+        return True
+
+    def _update_record(self, identifier, rtype=None, name=None, content=None):
```
Comment (adferrand) [inline lexicon/providers/safedns.py]: There is one case you did not implement here, which is related to the integration test `test_provider_when_calling_update_record_should_modify_record_name_specified`. It is when Lexicon is invoked to update a record, but no identifier is passed. In this case, the `--name` flag must be passed, and `name` and `rtype` are used together to find the record.

I usually reuse ̀self.list_records` for that. If there is no record found, then there is no update. If at least one record is found, the first one is updated (with a warning if there is more than one record).

Once you have implemented that, don't forget to remove `test_provider_when_calling_update_record_should_modify_record_name_specified` from the skipped tests, and fill its associated recording.

File: lexicon/providers/safedns.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        for delete_record_id in delete_record_ids:
+            self._delete(
+                '/zones/{0}/records/{1}'.format(self.domain, delete_record_id))
+
+        LOGGER.debug('delete_record: %s', True)
+        return True
+
+    # As there is no _patch method available from Provider, let's create our own
+    # by creating a _request with action 'PATCH'.
+    def _patch(self, url='/', data=None, query_params=None):
```
Comment (adferrand) [inline lexicon/providers/safedns.py]: I think you can move this method to the base provider, so everyone can benefit from it.

File: lexicon/providers/safedns.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    def _add_quotes(rtype, content):
+        if rtype == 'TXT':
+            if not content.startswith('"') and not content.endswith('"'):
+                return '"{0}"'.format(content)
+        return content
+
+    # When we're getting a response, we don't want quotes to wrap the content.
+    # This method removes them
+    @staticmethod
+    def _remove_quotes(rtype, content):
```
Comment (adferrand) [inline lexicon/providers/safedns.py]: Maybe you can reuse `_clean_TXT_record` from the base provider instead ?

File: lexicon/tests/providers/test_safedns.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    @pytest.mark.skip(reason="Record-level TTLs are not supported by this provider")
+    def test_provider_when_calling_list_records_after_setting_ttl(self):
+        return
+
+    @pytest.mark.skip(reason="CNAME requires FQDN for this provider")
+    def test_provider_when_calling_create_record_for_CNAME_with_valid_name_and_content(self):
+        return
+
+    @pytest.mark.skip(reason="new test, missing recording")
```
Comment (adferrand) [inline lexicon/tests/providers/test_safedns.py]: See my comment on `update_record` function in the provider.

File: lexicon/tests/providers/test_safedns.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    def test_provider_when_calling_update_record_should_modify_record_name_specified(self):
+        return
+
+    def _filter_response(self, response):
```
Comment (adferrand) [inline lexicon/tests/providers/test_safedns.py]: This override is not necessary.

File: lexicon/providers/safedns.py (hunk 2 above)
Comment (miff2000) [reply lexicon/providers/safedns.py]: Yep, good shout

File: lexicon/providers/safedns.py (hunk 3 above)
Comment (miff2000) [reply lexicon/providers/safedns.py]: I didn't spot that in the base class. Definitely!

File: lexicon/tests/providers/test_safedns.py (hunk 5 above)
Comment (miff2000) [reply lexicon/tests/providers/test_safedns.py]: Removed

File: lexicon/providers/safedns.py (hunk 1 above)
Comment (miff2000) [reply lexicon/providers/safedns.py]: I saw your logic in the ovh provider and I've mirrored similar in mine. Hopefully it satisfies.

**Comment by miff2000** [comment]
One question @adferrand. In my setup I find that the test `test_provider_when_calling_update_record_should_modify_record` was causing duplicate TXT records to be created. Did you find that also? I've added logic now to silently fail if that happens. Is that the right behaviour?

**Comment by adferrand** [comment]
> One question @adferrand. In my setup I find that the test `test_provider_when_calling_update_record_should_modify_record` was causing duplicate TXT records to be created. Did you find that also? I've added logic now to silently fail if that happens. Is that the right behaviour?

In fact you see a duplicate record because you launched (at least) two times the integration tests, without removing the created records between two consecutive iterations. Yes, these integration tests do not clean up things after execution, this is really something we should implement. This is not that obvious however, since if the clean up is relying on `delete_record` is broken at some point, you could end with a ineffective cleanup, or worse, an accidental wipe up of all the zone...

Nonetheless, you are seeing the records created by the previous iterations. So it is not necessary to put a condition for that. It may be counterproductive in fact, because it will make update_record return True, while no update has been done.

**Comment by miff2000** [comment]
> > One question @adferrand. In my setup I find that the test `test_provider_when_calling_update_record_should_modify_record` was causing duplicate TXT records to be created. Did you find that also? I've added logic now to silently fail if that happens. Is that the right behaviour?
> 
> In fact you see a duplicate record because you launched (at least) two times the integration tests, without removing the created records between two consecutive iterations. Yes, these integration tests do not clean up things after execution, this is really something we should implement. This is not that obvious however, since if the clean up is relying on `delete_record` and it is broken at some point, you could end with an ineffective cleanup, or worse, an accidental wipe of all the zone...
> 
> Nonetheless, you are seeing the records created by the previous iterations. So it is not necessary to put a condition for that and ignore it. It may be counterproductive in fact, because it will make update_record return True, while no update has been done.

Oh okay, that explains it. I wrongly assumed that the tests were intended to be able to be run without manual cleanup. So, with that in mind, this is the logic that I've borrowed (😉) from your OVH provider:
```python
            records = self._list_records(rtype, name)
            if len(records) == 1:
                identifier = records[0]['id']
            elif len(records) > 1:
                raise Exception('Several record identifiers match the request')
            else:
                raise Exception('Record identifier could not be found')
```
In your comment [above](https://github.com/AnalogJ/lexicon/pull/411#discussion_r294514983) you said to update the first record and warn. The two differ of course, so would you recommend that I update I raise an exception on the 0 records found, and update the first only and output a warning to LOGGER.debug if more >1 records found?

**Comment by miff2000** [comment]
Hey @adferrand, did you get chance to review these changes I made?

**Comment by adferrand** [comment]
Doing it today!

**Comment by adferrand** [comment]
About your question, indeed the behavior is not consistent between the providers. Including the ones I implemented, since we did not cleared that point in the spec. More generally, there were some discussions to make a more consistent API around update operations.

Since most recent providers update first matching record with a warning if there is more than one, I would do the update instead of raising an exception.

**Comment by miff2000** [comment]
> About your question, indeed the behavior is not consistent between the providers. Including the ones I implemented, since we did not cleared that point in the spec. More generally, there were some discussions to make a more consistent API around update operations.
> 
> Since most recent providers update first matching record with a warning if there is more than one, I would do the update instead of raising an exception.

Okay great, thanks. I'm switching to log instead as you've suggested and I'll push the changes as soon as I'm done. Shouldn't be long

**Comment by miff2000** [comment]
Pushed @adferrand and all CI steps except AppVeyor have passed. Hopefully we're ready to merge! 🙂

File: lexicon/providers/safedns.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        return True
+
+    def _update_record(self, identifier, rtype=None, name=None, content=None):
+
+        # Make sure the update won't cause a duplicate entry. If it will, fail silently
+        records = self._list_records(rtype, name, content)
+        if len(records) >= 1:
+            LOGGER.debug('update_record: (ignored, update would cause a duplicate record): %s',
+                         records[0]['id'])
+            return True
```
Comment (adferrand) [inline lexicon/providers/safedns.py]: Just one last thing, I think you can remove this logic now.

Comment (miff2000) [reply lexicon/providers/safedns.py]: Yes, annoyed I missed that. Corrected and pushed

**Pull request merged**